docs/proposals/runtime-client-server.md

```diff
@@ -0,0 +1,143 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest release of this document can be found
+[here](http://releases.k8s.io/release-1.1/docs/proposals/runtime-client-server.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Client/server implementation of container runtime
+
+## Abstract
+
+A proposal of client/server implementation of container runtime.
+
+## Motivation
+
+Currently, any container runtime has to be linked into the kubelet. This makes experimentation difficult, and prevents users from landing an alternate container runtime without landing code in core kubernetes.
+
+To facilitate experimentation and to enable user choice, this proposal adds a client/server implementation of the container runtime interface.
+
+This implementation will simply encode the requests, send them to a server where they will be decoded and sent into an instance of the container runtime interface.
+
+However, this enables container runtime implementations to be built and maintained outside of the core kubernetes tree.
+
+## Proposed design
+
+**Design of client/server container runtime**
+
+The main idea of client/server container runtime is to keep main control logic in kubelet while letting remote runtime only do dedicated actions:
+
+* Keep `SyncPod()` in kubelet and remote runtime only needs to start or kill a pod
+* Keep volume manager in kubelet and pass volumeMap to remote runtime when starting a new pod
+* Keep image garbage collection in kubelet and remote runtime only needs to pull or remove images
+
+Furthermore, kubelet network plugin and cAdvisor should make sense of container runtime type:
+
+* Kubelet network plugin has no sense of dedicated container runtime currently, but new container runtime may not work with existing docker or rkt plugin. The type of container runtime will be added to kubelet network plugin interface.
+* cAdvisor only supports docker and it is not working with other container runtimes. The type of container runtime will be added to cAdvisor Interface. Each new container runtime should add their own driver to cAdvisor.
+
+**`RemoteRuntime` Interface**
+
+```go
+type RemoteRuntime interface {
+  Type() string
+  Version() (Version, error)
+
+  // Pod APIs
+  GetPods(all bool) ([]*Pod, error)
+  GarbageCollect(gcPolicy ContainerGCPolicy) error
+  KillPod(pod *api.Pod, runningPod Pod) error
+  RunPod(pod *api.Pod, volumes kubecontainer.VolumeMap, options []kubecontainer.RunContainerOptions) error
```

Comment:
Can we only pass api.PodSpec here, and return PodStatus for both RunPod and KillPod? The initial reason of pasing entire api.Pod is that we have SyncPod method exposed at Container runtime, and engineers want to keep them consistent at params. If we decided to remove SyncPod from API, the restriction is not required anymore.